service/src/ConstConfigIOGroup.hpp

```diff
@@ -0,0 +1,115 @@
+            ///
+            /// @throws geopm::Exception there are no controls supported by the ConstConfigIOGroup
+            void write_control(const std::string &control_name, int domain_type, int domain_idx, double setting) override;
+            /// @brief Does nothing; this IOGroup does not provide any controls.
+            void save_control(void) override;
+            /// @brief Does nothing; this IOGroup does not provide any controls.
+            void restore_control(void) override;
+            std::function<double(const std::vector<double> &)> agg_function(const std::string &signal_name) const override;
+            std::function<std::string(double)> format_function(const std::string &signal_name) const override;
+            std::string signal_description(const std::string &signal_name) const override;
+            /// @throws geopm::Exception there are no controls supported by the ConstConfigIOGroup
+            std::string control_description(const std::string &control_name) const override;
+            /// @return M_SIGNAL_BEHAVIOR_CONSTANT 
+            int signal_behavior(const std::string &signal_name) const override;
+            /// @brief Does nothing; this IOGroup does not provide any controls.
+            ///
+            /// @param save_path this argument is ignored
+            void save_control(const std::string &save_path) override;
+            /// @brief Does nothing; this IOGroup does not provide any controls.
+            ///
+            /// @param save_path this argument is ignored
+            void restore_control(const std::string &save_path) override;
+            std::string name(void) const override;
+            static std::string plugin_name(void);
+            static std::unique_ptr<IOGroup> make_plugin(void);
+
+        private:
+            struct m_signal_desc_s {
+                json11::Json::Type type;
+                bool required;
+            };
+
+            struct m_signal_info_s {
+                int units;
+                int domain;
+                std::function<double(const std::vector<double> &)> agg_function;
+                std::string description;
+                std::vector<double> values;
+            };
+
+            struct m_signal_ref_s {
+                std::shared_ptr<m_signal_info_s> signal_info;
+                int domain_idx;
+            };
+
+            void parse_config_json(const std::string &config);
+            static json11::Json construct_config_json_obj(const std::string &config);
+            static void check_json_root(const json11::Json &root);
+            static void check_json_signal(const json11::Json::object::value_type &signal_obj);
+            static void check_json_signal_properties(const std::string &signal_name,
+                                                     const json11::Json::object& properties);
+            static void check_json_array_value(const json11::Json& val);
```

Comment:
Let's talk about what all we really need to expose as a static function at this time.

Reply:
What do you mean? These functions are private and not being exposed.